### PR TITLE
Implement ipNetToMediaPhysAddress (ARP table)

### PR DESCRIPTION
**Pull request by qiluo-msft**

### CODE REVIEW

File: src/ax_interface/encodings.py [hunk 1]
```diff
@@ -143,12 +143,12 @@ def size(self):
         return 4 + self.length + util.pad4(self.length)
 
     def __str__(self):
-        return self.string.decode('ascii')
+        return self.string.decode('utf-8')
```
Comment (pavel-shirshov) [inline src/ax_interface/encodings.py]: Why do we need this?

File: src/ax_interface/util.py [hunk 2]
```diff
@@ -98,3 +98,10 @@ def mac_decimals(mac):
     """
     return tuple(int(h, 16) for h in mac.split(":"))
 
+def ip2tuple(ip):
```
Comment (pavel-shirshov) [inline src/ax_interface/util.py]: ipv4 only? probably it's better to use ip2tuple_v4

File: src/sonic_ax_impl/mibs/__init__.py [hunk 3]
```diff
@@ -34,7 +34,14 @@ def get_index(if_name):
...
     """
-    match = re.match(SONIC_ETHERNET_RE_PATTERN, if_name.decode())
+    return get_index_from_str(if_name.decode())
+
+def get_index_from_str(if_name):
+    """
+    OIDs are 1-based, interfaces are 0-based, return the 1-based index
+    Ethernet N = N + 1
+    """
+    match = re.match(SONIC_ETHERNET_RE_PATTERN, if_name)
```
Comment (pavel-shirshov) [inline src/sonic_ax_impl/mibs/__init__.py]: Probably it's better to compile this regexp in __init__?
Also is it possible to extract this value from string using something like
if_name.replace('Ethernet', '') ?

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 4]
```diff
@@ -41,6 +44,53 @@ class DbTables(int, Enum):
...
+        for entry in python_arptable.get_arp_table():
+            dev = entry['Device']
+            mac = entry['HW address']
+            ip = entry['IP address']
+
+            if_index = mibs.get_index_from_str(dev)
+            if if_index is None: continue
+
+            mactuple = mac_decimals(mac)
+            machex = ''.join([chr(b) for b in mactuple])
```
Comment (pavel-shirshov) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: machex = ''.join(chr(b) for b in mactuple) should work for you

File: src/ax_interface/encodings.py (hunk 1 above)
Comment (qiluo-msft) [inline src/ax_interface/encodings.py]: Fix it and add test case.

---
In reply to: [108312790](https://github.com/Azure/sonic-snmpagent/pull/19#discussion_r108312790) [](ancestors = 108312790)

File: src/sonic_ax_impl/mibs/__init__.py (hunk 3 above)
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/__init__.py]: Let's keep it for readability and optimize if we find it on critical path. re library has the opportunity to hide the optimization without users changing code.

---
In reply to: [108313222](https://github.com/Azure/sonic-snmpagent/pull/19#discussion_r108313222) [](ancestors = 108313222)

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 5]
```diff
@@ -41,6 +44,53 @@ class DbTables(int, Enum):
...
+            mac = entry['HW address']
+            ip = entry['IP address']
+
+            if_index = mibs.get_index_from_str(dev)
+            if if_index is None: continue
+
+            mactuple = mac_decimals(mac)
+            machex = ''.join(chr(b) for b in mactuple)
+            # if MAC is all zero
+            #if not any(mac): continue
```
Comment (pavel-shirshov) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: Should we remove this line? What to do if the mac is 0?

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 6]
```diff
@@ -41,6 +44,53 @@ class DbTables(int, Enum):
...
+
+            iptuple = ip2tuple_v4(ip)
+
+            subid = (if_index,) + iptuple
+            self.arp_dest_map[subid] = machex
+            self.arp_dest_list.append(subid)
+
+            # print(subid, dev, mac, ip)
```
Comment (pavel-shirshov) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: Should we remove it?

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py (hunk 5 above)
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: The correct behavior of incomplete ARP entry (an IP without MAC) is not clear in the protocol. ref: [http://cric.grenoble.cnrs.fr/Administrateurs/Outils/MIBS/?oid=1.3.6.1.2.1.4.22.1.2](http://cric.grenoble.cnrs.fr/Administrateurs/Outils/MIBS/?oid=1.3.6.1.2.1.4.22.1.2)
Just keep the code here for further refinement, if needed.

---
In reply to: [108499363](https://github.com/Azure/sonic-snmpagent/pull/19#discussion_r108499363) [](ancestors = 108499363)

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py (hunk 6 above)
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py]: Removed

---
In reply to: [108499446](https://github.com/Azure/sonic-snmpagent/pull/19#discussion_r108499446) [](ancestors = 108499446)

**Pull request merged**